EX1/HW5.py

```diff
@@ -0,0 +1,118 @@
+import os
+import datetime
+import random
+import re
+
+
+
+def get_files_in_directory(path):
+    file_names = []
+    with os.scandir(path) as directory:
+        for entry in directory:
+            if entry.is_file() and entry.name.startswith("deep"):
+                file_names.append(os.path.basename(entry.path))
+    return file_names
+
+
+def random_date_between_dates(start_date, end_date):
+    start = datetime.datetime.strptime(start_date, '%Y-%m-%d')
+    end = datetime.datetime.strptime(end_date, '%Y-%m-%d')
+    delta = end - start
+    return (start + datetime.timedelta(days=random.randint(0, delta.days))).strftime("%A, %B %d, %Y")
+
+
+def join(*args, sep='-'):
+    result = []
+    for i, lst in enumerate(args):
+        result.extend(lst)
+        if i < len(args) - 1 and sep:
+            result.append(sep)
+    return result
+
+
+def get_recipe_price(prices, optionals=None, **ingredients):
+    total_price = 0
+    for ingredient, amount in ingredients.items():
+        if ingredient not in prices:
+            raise ValueError(f"{ingredient} is not a valid ingredient.")
+        if optionals and ingredient in optionals:
+            continue
+        price_per_100g = prices[ingredient]
+        ingredient_price = (price_per_100g / 100) * amount
+        total_price += ingredient_price
+    return round(total_price, 2)
+
+
+def perfect_numbers():
+    """Generate all perfect numbers."""
+    n = 1
+    while True:
+        divisors = [i for i in range(1, n) if n % i == 0]
+        if sum(divisors) == n:
+            yield n
+        n += 1
+
+
+def extract_messages(filename):  # TODO  fix it.
+    """Extract hidden messages from a binary file."""
+    msg_buffer = ''
+    with open('resources/logo.jpg', 'rb') as f:
```

Comment:
A good rule of thumb is not to have more than 3 tabs, you could use several solutions. 

1. Separate code to small functions- 
https://softwareengineering.stackexchange.com/questions/308108/when-is-it-appropriate-to-make-a-separate-function-when-there-will-only-ever-be

2. Replace for loops -  
https://medium.com/python-pandemonium/never-write-for-loops-again-91a5a4c84baf
Replace for loops with more advanced python features 

3.  You should aim to write everything in a single line, even if you can. 
This link https://towardsdatascience.com/you-should-stop-craving-for-python-one-liners-5dbf8e0de23d explains it well, even though I disagree with how important memory and runtime are in most python programs.